internal/api/callbacks.go

```diff
@@ -263,8 +263,12 @@ func cScan(ptr *C.db_t, gasMeter *C.gas_meter_t, usedGas *cu64, start C.U8SliceV
 		return C.GoError_User
 	}
 
-	out.state = cIterator
-	out.vtable = iterator_vtable
+	*out = C.GoIter{
```

Comment:
This line seems to cause a "Go pointer stored into non-Go memory" error which I do not yet understand very well

Reply:
Okay I found it and it's interesting. The problem was not `C.GoIter{` but the gasMeter pointer. The gasMeter originally is a Go variable under the control of the GC. Now the checker `cgocheck2` can detect if a pointer to Go memory is stored outside of Go, potentially leading to situations where the pointer stored in C is used after the object was freed or moved. [The Pinner API ](https://pkg.go.dev/runtime#Pinner)is designed to tell the GC: please don't free or move. This was[ done here](868c5a70cbe0ba695f232a7fa9c4c240a217639b), avoiding the previous error message without further tricks. This change makes a lot of sense, especially since the gasMeter is stored in many places in the Rust code, such as Db.

I'm a bit surprised this did not pop up before but maybe it did and this is why the code set the gas meter in Rust before. For other variables we might be hiding the problem through the use of `unsafe.Pointer`.

Reply:
Interesting, and `runtime.KeepAlive` would not work for that?

Reply:
I am not filly clear about the difference as I e.g. don't really understand what a finalizer in Go is. But since the cgocheck2 check explicitly checks for pinning, this seems to be the right way.

Reply:
From what I understand, the `KeepAlive` only prevents garbage collection, while the pinner prevents moving in addition to that. I am not sure either which one we would need here, so better to err on the safe side.